# HG changeset patch
# User Rik <octave@nomad.inbox5.com>
# Date 1307196942 25200
#      Sat Jun 04 07:15:42 2011 -0700
# Branch stable
# Node ID 9843b3b055e094586e442b57b30cb2a8f7bc1738
# Parent  fb93b94dfc8260b9e02dd26033f191306a26345b
Fix bug with initial conditions in filter (Bug #32741).

* filter.cc: Re-orient SI to column vector only when both SI and X are vectors.

diff --git a/src/DLD-FUNCTIONS/filter.cc b/src/DLD-FUNCTIONS/filter.cc
--- a/src/DLD-FUNCTIONS/filter.cc
+++ b/src/DLD-FUNCTIONS/filter.cc
@@ -451,29 +451,20 @@ H(z) = {\\displaystyle\\sum_{k=0}^M d_{k
                   for (int i = dim; i > 0; i--)
                     si_dims(i) = si_dims(i-1);
                   si_dims(0) = si_len;
 
                   si.resize (si_dims, 0.0);
                 }
               else
                 {
-                  dim_vector si_dims = args (3).dims ();
-                  bool si_is_vector = true;
-                  for (int i = 0; i < si_dims.length (); i++)
-                    if (si_dims(i) != 1 && si_dims(i) < si_dims.numel ())
-                      {
-                        si_is_vector = false;
-                        break;
-                      }
-
                   si = args(3).float_complex_array_value ();
 
-                  if (si_is_vector)
-                    si = si.reshape (dim_vector (1, si.numel ()));
+                  if (si.is_vector () && x.is_vector ())
+                    si = si.reshape (dim_vector (si.numel (), 1));
                 }
 
               if (! error_state)
                 {
                   FloatComplexNDArray y (filter (b, a, x, si, dim));
 
                   if (nargout == 2)
                     retval(1) = si;
@@ -508,29 +499,20 @@ H(z) = {\\displaystyle\\sum_{k=0}^M d_{k
                   for (int i = dim; i > 0; i--)
                     si_dims(i) = si_dims(i-1);
                   si_dims(0) = si_len;
 
                   si.resize (si_dims, 0.0);
                 }
               else
                 {
-                  dim_vector si_dims = args (3).dims ();
-                  bool si_is_vector = true;
-                  for (int i = 0; i < si_dims.length (); i++)
-                    if (si_dims(i) != 1 && si_dims(i) < si_dims.numel ())
-                      {
-                        si_is_vector = false;
-                        break;
-                      }
-
                   si = args(3).complex_array_value ();
 
-                  if (si_is_vector)
-                    si = si.reshape (dim_vector (1, si.numel ()));
+                  if (si.is_vector () && x.is_vector ())
+                    si = si.reshape (dim_vector (si.numel (), 1));
                 }
 
               if (! error_state)
                 {
                   ComplexNDArray y (filter (b, a, x, si, dim));
 
                   if (nargout == 2)
                     retval(1) = si;
@@ -568,29 +550,20 @@ H(z) = {\\displaystyle\\sum_{k=0}^M d_{k
                   for (int i = dim; i > 0; i--)
                     si_dims(i) = si_dims(i-1);
                   si_dims(0) = si_len;
 
                   si.resize (si_dims, 0.0);
                 }
               else
                 {
-                  dim_vector si_dims = args (3).dims ();
-                  bool si_is_vector = true;
-                  for (int i = 0; i < si_dims.length (); i++)
-                    if (si_dims(i) != 1 && si_dims(i) < si_dims.numel ())
-                      {
-                        si_is_vector = false;
-                        break;
-                      }
-
                   si = args(3).float_array_value ();
 
-                  if (si_is_vector)
-                    si = si.reshape (dim_vector (1, si.numel ()));
+                  if (si.is_vector () && x.is_vector ())
+                    si = si.reshape (dim_vector (si.numel (), 1));
                 }
 
               if (! error_state)
                 {
                   FloatNDArray y (filter (b, a, x, si, dim));
 
                   if (nargout == 2)
                     retval(1) = si;
@@ -625,29 +598,20 @@ H(z) = {\\displaystyle\\sum_{k=0}^M d_{k
                   for (int i = dim; i > 0; i--)
                     si_dims(i) = si_dims(i-1);
                   si_dims(0) = si_len;
 
                   si.resize (si_dims, 0.0);
                 }
               else
                 {
-                  dim_vector si_dims = args (3).dims ();
-                  bool si_is_vector = true;
-                  for (int i = 0; i < si_dims.length (); i++)
-                    if (si_dims(i) != 1 && si_dims(i) < si_dims.numel ())
-                      {
-                        si_is_vector = false;
-                        break;
-                      }
-
                   si = args(3).array_value ();
 
-                  if (si_is_vector)
-                    si = si.reshape (dim_vector (1, si.numel ()));
+                  if (si.is_vector () && x.is_vector ())
+                    si = si.reshape (dim_vector (si.numel (), 1));
                 }
 
               if (! error_state)
                 {
                   NDArray y (filter (b, a, x, si, dim));
 
                   if (nargout == 2)
                     retval(1) = si;
@@ -744,13 +708,18 @@ filter (MArray<FloatComplex>&, MArray<Fl
 %!  a = [1,1]; b=[1,1];
 %!  x = zeros (4,4,2); x(1,1:4,1) = +1; x(1,1:4,2) = -1;
 %!  y0 = zeros (4,4,2); y0(1:2,1:4,1) = +1; y0(1:2,1:4,2) = -1;
 %!  y = filter (b, [1], x);
 %!  assert(y, y0);
 %!
 %!assert(filter (1, ones(10,1)/10, []), []);
 %!assert(filter (1, ones(10,1)/10, zeros(0,10)), zeros(0,10));
+%!assert(filter (1, ones(10,1)/10, single (1:5)), repmat (single (10), 1, 5));
+%% Test using initial conditions
+%!assert(filter([1, 1, 1], [1, 1], [1 2], [1, 1]), [2 2]);
+%!assert(filter([1, 1, 1], [1, 1], [1 2], [1, 1]'), [2 2]);
 %!assert(filter([1, 3], [1], [1 2; 3 4; 5 6], [4, 5]), [5 7; 6 10; 14 18]);
-%!assert(filter (1, ones(10,1)/10, single (1:5)), repmat (single (10), 1, 5));
+%!error (filter([1, 3], [1], [1 2; 3 4; 5 6], [4, 5]'));
+%!assert(filter([1, 3, 2], [1], [1 2; 3 4; 5 6], [1 0 0; 1 0 0], 2), [2 6; 3 13; 5 21]);
 %%  Should put some tests of the "DIM" parameter in here.
 
 */
